fix(a11y): settings toggles

bbsmooth opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #49214 

<!-- Feel free to add any additional description of changes below this line -->

Sorry @Sboonny, I didn't realize you were working on this as well.

I just went with the browser's default radio buttons. Not sure if we want to use customized buttons?

Sboonny commented:
> Sorry @Sboonny, I didn't realize you were working on this as well.

Didn't do much, to be honest, and I am always happy with the help, ping me if you want me to pick up a task.

bbsmooth committed:
feat: add keyboard focus for radio buttons

bbsmooth commented:
@Sboonny What I really need to know is how we want to style the radio buttons. That's the main blocker.

Sboonny commented:
I don't mind the current look, but I will bring this to the Wednesday meeting, if needed 👍

bbsmooth committed:
fix: focus outline should show on all four sides of button

ahmadabdolsaheb commented:
We could go with something like this for the styles:
<img width="1638" alt="Screenshot 2023-03-15 at 12 23 34 PM" src="https://user-images.githubusercontent.com/4591597/225265602-a857682a-bc71-4345-821a-cf9b2520f954.png">

Also the tick for the toggle could come back if it does not have any accessibility issues.

bbsmooth commented:
@ahmadabdolsaheb I would be hesitant to use square boxes here since they are usually associated with check boxes, not radio buttons.

bbsmooth committed:
refactor: remove unnecessary ternary

bbsmooth committed:
fix: focused button should not stack on top of flash alert

bbsmooth committed:
remove toggle-settings.tsx

chore: remove unnecessary elements

remove unnecessary CSS/fix responsiveness

feat: add keyboard focus for radio buttons

fix: use defaultChecked and onChange for radio buttons

fix: do not change color on hover over pressed button

fix: focus outline should show on all four sides of button

chore: setPrivacyTogglesToPublic

feat: add tick mark to active button

refactor: remove unnecessary ternary

fix: focused button should not stack on top of flash alert

feat: standardize radio button color

fix: remove random id generation

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+  toggleFlag,
+  ...restProps
+}: ToggleRadioSettingProps): JSX.Element {
+  const randomId = Math.floor(Math.random() * 1_000_000);
+  const firstRadioId = `radioA${randomId}`;
+  const secondRadioId = `radioB${randomId}`;
```
Same as toggle button
```suggestion
```

bbsmooth replied:
I think it is best if we leave the `firstRadioId` and `secondRadioId` declarations here (and just use `flagName` instead of `randomId`) since these values are each used in two places below and thus if changed them for some reason we would only need to change them once here.

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+    <div className='toggle-setting-container'>
+      <fieldset
+        {...(explain && {
+          'aria-labelledby': `legend${randomId} desc${randomId}`
```
```suggestion
          'aria-labelledby': `legend${flagName} desc${flagName}`
```

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+      >
+        <legend
+          className='sr-only'
+          {...(explain && { id: `legend${randomId}` })}
```
```suggestion
          {...(explain && { id: `legend${flagName}` })}
```

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+        </legend>
+        <div className='toggle-description'>
+          <p aria-hidden={true}>{action}</p>
+          {explain ? <p id={`desc${randomId}`}>{explain}</p> : null}
```
```suggestion
          {explain ? <p id={`desc${flagName}`}>{explain}</p> : null}
```

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+          {explain ? <p id={`desc${randomId}`}>{explain}</p> : null}
+        </div>
+        <div className='toggle-radio-group'>
+          <label htmlFor={firstRadioId} data-checked={flag}>
```
```suggestion
          <label htmlFor={flagName} data-checked={flag}>
```

bbsmooth replied:
Leaving this as `firstRadioId` since it is used in more than one place.

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+        <div className='toggle-radio-group'>
+          <label htmlFor={firstRadioId} data-checked={flag}>
+            <input
+              id={firstRadioId}
```
```suggestion
              id={flagName}
```

bbsmooth replied:
Leaving this as `firstRadioId` since it is used in more than one place.

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+          </label>
+          <label htmlFor={secondRadioId} data-checked={!flag}>
+            <input
+              id={secondRadioId}
```
```suggestion
              id={flagName}
```

bbsmooth replied:
Leaving this as `secondRadioId` since it is used in more than one place.

Sboonny commented on client/src/components/settings/toggle-radio-setting.tsx:
```diff
@@ -0,0 +1,73 @@
+            />
+            <span>{restProps.onLabel}</span>
+          </label>
+          <label htmlFor={secondRadioId} data-checked={!flag}>
```
```suggestion
          <label htmlFor={flagName} data-checked={!flag}>
```

bbsmooth replied:
Leaving this as `secondRadioId` since it is used in more than one place.

Sboonny commented:
Just wondering, do we need aria-labelledby? I though legend with screen-only class is good enough to act as aria-labelledby.

What am I missing? https://www.w3.org/WAI/tutorials/forms/labels/#using-aria-labelledby

bbsmooth commented on client/src/components/settings/toggle-button-setting.tsx:
```diff
@@ -0,0 +1,88 @@
+  toggleFlag,
+  ...restProps
+}: ToggleButtonSettingProps): JSX.Element {
+  const randomId = Math.floor(Math.random() * 1_000_000);
```
Ahh, thanks for pointing this out. I think the reason I didn't use `flagName` in the first place is because originally they were not unique. If you look at the donation toggle you'll see that it was originally using `showPortfolio` for the `flagName` by mistake, thus there were duplicate `flagName` values. Even though I fixed that and they are all unique now, I must have forgot about it and decided to create a random ID instead to guarantee uniqueness.

Sboonny replied:
Didn't notice the comment until ahmad pushed it to the notification, sorry for late reply.

I think this is the wrong mindset to have, there is 1/100000 chance that the random function id will provide the same id, so instead of relaying on chance to provide unique description, we can make it more obvious that `flagName` should be unique:

- adding comments.
- changing the variable to `screenReaderContext`.
- providing `@example` and `@params` in TypeScript, to show as hints in the editor.

it isn't enforcing the unique `flagName`, but people will be going out of their way to ignore all of these hints, and if contributor do, members shouldn't.

---

I know that we are people that do make mistakes, I can say that because I am the most person to make mistakes, but the solution shouldn't be making an abstract solution that can be forgotten, then removed because someone didn't remember the importance of the uniqueness.

but we have to get better and do better, and provide better info through the code that this is something that should be respected for these reasons.

bbsmooth replied:
@Sboonny 
> there is 1/100000 chance that the random function id will provide the same id

I don't think this is an issue. React 18 now provides a `useId` hook which guarantees a unique id, so we can ultimately use that. In the meantime, there are modules we could use instead to get the same thing, such as react-id-generator. Or we could do our own checks to make sure the id generated by `random()` doesn't exist and just try again if it does. We could also use `Cryto.getRandomValues()` instead to get even better random numbers to decrease the chances of duplicates.

My point is that these ids are being used behind the scenes merely for accessibility reasons and thus they do not need to be known by the programmer, and should not depend on the programmer to give them a value since there is a chance of programmer error. I don't think this is an abstract solution. This is a perfectly reasonable solution to ensure that accessibility is met, and in my opinion, a much cleaner solution than all of the bullets points we'd have to implement above. And the reason for generating the id would not be forgotten as we can add a comment at the source explaining why we are generating a unique id.

But in reality, the chances of a developer using the same flagName is probably close to nil and so this is not a battle worth continuing :slightly_smiling_face: Because the accessibility of these controls needs urgent fixing and I would like to see it happen ASAP, I will acquiesce and make the suggested changes so we can get this merged.

Sboonny replied:
No need, if there's a better solution that provide a granted unique I'd, we can move to them later.

I was not aware of these other options q👍

bbsmooth commented:
> do we need aria-labelledby?

You'll notice that `aria-labelledby` is only used when the toggle has extra description text. In that case, we use `aria-labelledby` to change the accessible name to include both the toggle name and the description text so that screen readers will announce both when the user is interacting with the toggle. This is a standard best practice pattern.

ahmadabdolsaheb commented:
> @ahmadabdolsaheb I would be hesitant to use square boxes here since they are usually associated with check boxes, not radio buttons.
Sure, we could go with round ones.

bbsmooth committed:
fix: remove random id generation

Sboonny reviewed (approved):
Btw, you are right in this https://github.com/freeCodeCamp/freeCodeCamp/pull/49664#discussion_r1145562584, if you want to revert the lastest commit, please do so.

---

My suggestion was based on lack of knowledge, which you have provided. this is good to go after fixing the conflicts 👍.

Sboonny commented:
@bbsmooth, Is it okay if I push a commit that sort RTL, and fix the conflict?

bbsmooth commented:
@Sboonny Yes, go ahead.

Sboonny committed:
import type from toggleSetting to align the props

Commit:
display buttons as grids

Commit:
change value to string

The value is type string in the ui-component, and this aims to sort that

Commit:
clean a selector

Commit:
remove extra CSS properties

bbsmooth committed:
merge updates

bbsmooth committed:
remove duplicates

bbsmooth commented:
@Sboonny, are you done with your RTL updates? Everything looks pretty good, there are just a few minor things I'd like to adjust, but I want to wait until you are done.

Sboonny commented:
Yeab, please go ahead. 👍

bbsmooth committed:
tidy up styling on toggle buttons

bbsmooth committed:
tidy up styling on radio buttons

bbsmooth commented:
@Sboonny I updated the styling on the buttons so that the text is always centered and the check mark will be positioned just to the outside of the text. I also added a slight gap between the buttons so that the outline focus indicator would show on all four sides of both buttons. But if you would rather not have that `3px` gap then we can add a background color to the `.toggle-button-group` wrapper `div` to fill that space.

The radio buttons still look the same but I switched their various margins to logical properties so they would look the same in rtl languages.

Sboonny commented:
> But if you would rather not have that `3px` gap then we can add a background color to the `.toggle-button-group` wrapper `div` to fill that space. Or, we could remove the `3px` gap and then use relative positioning and `z-index` on the button when focused to make sure that it stacks above the other button so that all four sides of the outline show. 


No need, Idk what the others think, but I am happy with the current looks 👍.

ahmadabdolsaheb commented on client/src/components/settings/toggle-button-setting.tsx:
```diff
@@ -0,0 +1,88 @@
+  offLabel: string;
+  onLabel: string;
+};
+
```
The styles could be moved the a css file.

ahmadabdolsaheb commented:
Thank you for the hard work here.
 The difference between the inputs are harder to distinguish in dark mode than it is in Light mode.

<img width="1728" alt="Screenshot 2023-04-04 at 2 22 13 PM" src="https://user-images.githubusercontent.com/4591597/229782380-1a9172b0-c505-4957-846b-970141923dc6.png">

Would it be possible to set the border of the unselected input to the secondary-color and the background to transparent or the secondary-background?

bbsmooth commented:
@ahmadabdolsaheb I agree, in dark mode it is hard to tell the difference between the selected and non-selected radio button. Browsers don't provide a lot of options for styling native radio buttons. I will see if I can somehow make the difference between the two states larger using only CSS but I fear I we may need to switch to some sort of custom solution.

bbsmooth committed:
custom radio button styling

bbsmooth commented:
@ahmadabdolsaheb Custom radio button styling added. Should be very clear now both in light and dark mode.

ahmaxed reviewed (approved):
Thanks for the hard here. I want to get this in asap, but there are a few changes that improve the toggle buttons in another pr:
- toggle-button.tsx in the helper folder could be removed since it is not used anywhere.
- checkIconStyle should be moved to a CSS file.
- we have toggle buttons in video courses (scientific-computing-with-python, etc.), the newly added toggles should replicate the style of the pre-exiting toggles for consistency.
